### DISCUSSION TITLE
How to use "css-has-pseudo" in build step?

### DISCUSSION THREAD

**Discussion by Annie972297** [csstools/postcss-plugins#1532] [Q&A] [answered]
Hi, I want to use "css-has-pseudo" to improve the compatibility of ":has()". I have noticed the usage in Brower in the README.md as below,and it works in browsers.
`import cssHasPseudo from '<packageName>/browser';
cssHasPseudo(document);`
But it doesn't work in my build step, and I don't find any description in the README.md. How can I use "css-has-pseudo" in build step?

**Comment by romainmenke** [comment]
Hi @Annie972297,

Have you seen my answer here: https://github.com/csstools/postcss-plugins/issues/1504#issuecomment-2462190273

**Reply by romainmenke** [reply]
Also please provide a reproduction.
A project I can run and inspect.
Or a link to your website.

**Reply by Annie972297** [reply]
Hi, I have seen this. Sorry that I didn't make a prompt reply that time~
I am using Vite for run and build step. There's no error in the browser console. When I run the project locally with a mock server, it's ok.
But when I build the project, the css files are ok (which means there're "[csstools-has-xxx" in them), but there's no "[csstools-has-xxx" attribute added to the dom elements. I think only when the dom elements have  "[csstools-has-xxx" attribute can they have corresponding styles.

**Reply by romainmenke** [reply]
I do believe you, unfortunately I can not say anything particularly helpful without having something to debug.

Can you share anything of your project?
A repo, a public url.

If this is a private project we could also work something out :)

**Reply by Annie972297** [reply]
Hi, I upload a repo, please have a check: https://github.com/Annie972297/css-has-test.
Is my question about the enviroment? My project is using a device as server and I notice that it says "PostCSS Has Pseudo runs in all Node environments" in README.

**Reply by romainmenke** [reply]
Thank you for providing that extra information @Annie972297

Unfortunately I didn't see anything wrong.
As far as I could tell the polyfill worked on all devices that needed it :/

Can you provide more information about this step:

> But if I build the project and put it in a real device, it doesn't work. (My project is using a router as a server.)

- how do you build the project?
- which device?
- what does "using a router as a server" mean?

I don't know anything about vite, vue, ... so I can't help there.
If the issue can't be replicated on my end I won't be able to help :)

A screen capture might also provide some details we are overlooking?

**Reply by Annie972297** [reply]
Hi, I build the project just using "pnpm run build" and in fact vite automatically builds it. My project is designed to work in a wireless router as the administration page.
In the attached image, the dom has a "[csstools-has-xxx" attribute. When I access the page of the router, there's no "[csstools-has-xxx" attributes on elements. I don't know why.
I also download the original code of the plugin and try to find how does it add "[csstools-has-xxx" attributes to the elements. Do you know about that?
![csstools-has-xxx attribute](https://github.com/user-attachments/assets/11371aee-e182-4137-a775-856096fe777d)

**Reply by romainmenke** [reply]
There are two parts:
- the postcss plugin which rewrites your CSS
- the JavaScript polyfill which adds `csstools-has-..` attributes to elements on a page

The JavaScript polyfills runs clients side, in the browser. Not on the server or during the build.

----

To debug any issues with this plugin you should:
- check if the CSS contains rules with `csstools-has-...` selectors
- check if the elements on the page receive `csstools-has-` attributes

This plugin and polyfill are extremely well tested and are known to work.
So if anything is going wrong it must be caused by your setup.

It is perfectly possibly that your setup or framework is breaking this feature :/

**Reply by romainmenke** [reply]
Some things which might be causing issues:

- some other process is also changing your CSS selectors (css modules, scoped css, ... ?)
- the JavaScript polyfill is not loaded
- the JavaScript polyfill is throwing errors (run it with `cssHasPseudo(document, { debug: true });`)
- the CSS stylesheet can't be parsed by the JavaScript polyfill

**Reply by Annie972297** [reply]
There're `csstools-has-` selectors in CSS. So I think the problem is that the elements don't receive `csstools-has-` attributes?
Please check the images. I think something in the build step affects the polyfill? Do you know where is the code about the polyfill?
 
run locally:
![dom when run locally](https://github.com/user-attachments/assets/1b4295cf-7fdb-41d4-ab48-221549eee73f)

after build:
![css and dom](https://github.com/user-attachments/assets/6eee9343-3a3d-434e-a861-213fc4b76724)

**Reply by romainmenke** [reply]
Did you already run the polyfill with `cssHasPseudo(document, { debug: true })`?
Did that reveal any errors?

-----

Also keep in mind that the JavaScript polyfill doesn't run at all if your browser supports `:has()`. So in that case you won't see the attributes.

----

Is it possible that the `data-v-...` attributes are different?
The `:has()` polyfill encodes the entire selector.

-----

Can you try with unscoped styles?

I don't know anything about Vue, vite, ... so I can't tell you how to write unscoped styles.
But I suspect that the `[data-v-....]` selectors might be breaking the polyfill.

**Reply by Annie972297** [reply]
Hi, sorry that I have edited my last comment twice. I'm testing on Firefox 110.0 which doesn't support :has().
When running locally, it's ok, shown in the first image in my last comment. But after build, the elements can't receive `csstools-has-` attributes. 
I also try to run it with `cssHasPseudo(document, { debug: true })`, the Firefox 110.0 console says "DOMException: Document.querySelectorAll: '' is not a valid selector". But the plugin works and the style on my webpage is ok.
![run with debug true](https://github.com/user-attachments/assets/df038bab-c227-400d-84ff-d93d335b9285)


So I think the build step affects the polyfill? Do you know where is the code about the polyfill or how does it work?

**Reply by romainmenke** [reply]
All the code is here: https://github.com/csstools/postcss-plugins/tree/main/plugins/css-has-pseudo/src

The plugin and polyfill work by encoding all selectors that have a `:has()` selector in them.

`.foo :has(.bar)` -> `[csstools-has-...]`

The JavaScript polyfill then scans your stylesheet for such attribute selectors and decodes these.
It then uses the decoded selector as instructions to scan your document.
When it finds elements that would have matched the original selector it assigns the attribute selector to them.

The encoding and decoding is similar to base64 encoding, but specialized to be valid selectors.

**Reply by Annie972297** [reply]
Hi, I add two breakpoints in the function `cssHasPseudo` to debug: the first one is on `[].forEach.call(document.styleSheets, walkStyleSheet);`, and the second one is in the function `walkStyleSheet`. I also add `console.log(document.styleSheets)` just before the first breakpoint.
After build, one of page uses two stylesheets, one is for common (A.css) and the other one is only for this page (B.css). From the console, I see that the debug code `console.log(document.styleSheets)` only prints out "A.css". When the webpage breaks at the breakpoints, the browser hasn't request "B.css" yet. So it seems that the plugin doesn't `walkStyleSheet` "B.css".
It's a little weird because it works fine in development mode.

**Reply by romainmenke** [reply]
That is interesting.
Can you update your reproduction to also have these two stylesheets?

**Reply by romainmenke** [reply]
Also, the more I hear about this issue, the more sure I am that this is an issue in vite/vue and doesn't have anything to do with the PostCSS plugin.

So it might also be good to open an issue there?
They might be able to help you more quickly.

**Reply by Annie972297** [reply]
Hi, I find the key to reproduce my problem and update my repo, please have a check: https://github.com/Annie972297/css-has-test.

**Reply by Annie972297** [reply]
It seems that the plugin doesn't process async stylesheets? I haven't totally understand the principle of this plugin.

**Reply by Annie972297** [reply]
Hi, I create a MutationObserver in brower console to debug.
In the production mode, the async sheets aren't received by `MutationObserver`. They are imported by `<link href="...css">`. The mutationObserver I created notices the node, but can't get its sheet (node.sheet is `null`). While in the development mode, the sheets are imported by `<style>...</style>` (internal stylesheet). The mutationObserver can get its sheet.
Is there a method to improve the plugin?

The mutationObserver I created:
`var mutationObserver1 = new MutationObserver((mutationsList) => {
  mutationsList.forEach(mutation => {
    [].forEach.call(mutation.addedNodes || [], node => {
                console.log('node:', node, node.nodeType, node.sheet);
      // walk stylesheets to collect observed css rules
      if (node.nodeType === 1 && node.sheet) {
        console.log('nodeBingo:', node);
        console.log('node.sheet:', node.sheet);
      }
    });
  });
});
mutationObserver1.observe(document, { childList: true, subtree: true, attributes: true });`

![mutationObserverConsole](https://github.com/user-attachments/assets/5c85d119-6a86-44e0-bc26-9e1ac85c803c)

**Reply by romainmenke** [reply, marked as answer]
Hi @Annie972297,

Thank you so much for working on this repro.
I've made some changes and published a new version.

This new version now listens for the `load` event on stylesheets.
This should resolve your issue.

Can you try this version and let me know if this works?

**Reply by Annie972297** [reply]
Hi, I update the version to 7.0.2, it works! Thank you!
